File: examples/pulse-ext-example/java/src/main/java/com/inventzia/pulse/ext/examples/ExtendedBarRun.java
/*
 * SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Inventzia-Commercial
 * Copyright (c) 2013-2026 Magrino Bini, Paola Apruzzese, Inventzia Science and Technology Ltd.
 */
package com.inventzia.pulse.ext.examples;

import com.inventzia.pulse.beacon.core.AbstractActor;
import com.inventzia.pulse.beacon.core.AbstractGateway;
import com.inventzia.pulse.beacon.core.ComponentReporter;
import com.inventzia.pulse.beacon.core.Gateway;
import com.inventzia.pulse.beacon.core.GatewayStatus;
import com.inventzia.pulse.beacon.core.MultiClientEngine;
import com.inventzia.pulse.beacon.core.OperatingMode;
import com.inventzia.pulse.beacon.core.RunInfo;
import com.inventzia.pulse.beacon.core.Slf4jReporter;
import com.inventzia.pulse.beacon.core.Topic;
import com.inventzia.pulse.data.datum.Datum;
import com.inventzia.pulse.ext.schemas.ExtendedBar;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Shared run logic behind the two runnable examples,
 * {@link HistoricExtendedBarExample} and {@link RealTimeExtendedBarExample}: the extension's
 * {@link ExtendedBar} flowing through the engine, no Python involved.
 *
 * <p>An in-code source gateway emits a short stream of {@code ExtendedBar} (the ad-hoc datum
 * defined by this package), the engine routes it in event-time order, and a recording actor
 * receives it with every field intact. Because {@code ExtendedBar}'s jar carries a
 * {@code META-INF/services} entry for its {@code DatumTypeProvider}, the engine's registry
 * discovers it through the Service Provider Interface alongside the core provider, with no
 * change to pulse-data or pulse-beacon; the run logs both providers via {@link RunInfo}.
 *
 * <p>This class holds no {@code main}: run one of the two example classes instead. The only
 * difference between them is the time window, which selects {@link OperatingMode#COMPRESSED_TIME}
 * (historical replay) or {@link OperatingMode#REAL_TIME} (paced to the wall clock).
 */
final class ExtendedBarRun {

    private static final ComponentReporter LOG =
            new ComponentReporter("extended-bar-run", Slf4jReporter.shared());

    private static final String SYMB  = "AAPL";
    private static final String TOPIC = "ext.bars";

    /** op, hi, lo, cl, bidVolume, askVolume, tradeCount (mirrors the Python example's rows). */
    private record Bar(BigDecimal op, BigDecimal hi, BigDecimal lo, BigDecimal cl,
                       long bidVolume, long askVolume, long tradeCount) {}

    private static final List<Bar> BARS = List.of(
            new Bar(bd("1.00"), bd("1.20"), bd("0.95"), bd("1.10"), 40, 60, 3),
            new Bar(bd("1.10"), bd("1.30"), bd("1.05"), bd("1.25"), 55, 45, 5),
            new Bar(bd("1.25"), bd("1.28"), bd("1.10"), bd("1.15"), 30, 70, 4));

    private static BigDecimal bd(String s) { return new BigDecimal(s); }

    private static ExtendedBar bar(long t, Bar b) {
        return new ExtendedBar(
                SYMB, t, b.op(), b.hi(), b.lo(), b.cl(),
                BigDecimal.valueOf(b.bidVolume() + b.askVolume()),   // vlm
                BigDecimal.valueOf(b.bidVolume()),
                BigDecimal.valueOf(b.askVolume()),
                b.tradeCount(), "XNAS");
    }

    /** The rows we expect the consumer to record, in order (symb|time|cl|tradeCount). */
    private static List<String> expected(long[] times) {
        List<String> out = new ArrayList<>();
        for (int i = 0; i < times.length; i++) {
            Bar b = BARS.get(i);
            out.add(SYMB + "|" + times[i] + "|" + b.cl() + "|" + b.tradeCount());
        }
        return out;
    }

    // ------------------------------------------------------------------
    // A clock-driving source gateway that emits the fixed ExtendedBar stream.
    // Same shape as HeartBeatGateway: produces, never receives.
    // ------------------------------------------------------------------
    private static final class ExtendedBarFeed extends AbstractGateway {

        private final Topic<ExtendedBar> topic;
        private final long[] times;

        ExtendedBarFeed(String name, Topic<ExtendedBar> topic, long[] times, long start, long end) {
            super(name, start, end);
            this.topic = topic;
            this.times = times;
            setDriveClock(true);
        }

        @Override
        public void run() {
            initialize();
            connect();
            setStatus(GatewayStatus.STARTED);
            boolean realTime = operatingMode() == OperatingMode.REAL_TIME;
            try {
                for (int i = 0; i < times.length && connected(); i++) {
                    long t = times[i];
                    if (realTime) {
                        long wait = t - System.currentTimeMillis();
                        if (wait > 0) Thread.sleep(wait);
                        if (!connected()) break;
                    }
                    Gateway downstream = subscriberForKey(topic, SYMB);
                    if (downstream != null) {
                        // In compressed time this blocks on the time-machine permit until the
                        // previous bar is consumed, pacing the loop in event-time order.
                        downstream.onEvent(topic, bar(t, BARS.get(i)));
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            disconnect();
            setStatus(GatewayStatus.STOPPED);
        }

        @Override public <P extends Datum> void publish(Topic<P> topic, P payload) {
            throw new UnsupportedOperationException(name() + ": source gateway does not accept publish()");
        }

        @Override public <P extends Datum> void onEvent(Topic<P> topic, P payload) {
            throw new UnsupportedOperationException(name() + ": source gateway does not accept incoming events");
        }
    }

    // ------------------------------------------------------------------
    // A recording actor: keeps what it received and logs each bar's fields.
    // ------------------------------------------------------------------
    private static final class RecordingConsumer extends AbstractActor {

        private final List<String> received = new ArrayList<>();

        RecordingConsumer(String name) { super(name); }

        List<String> received() { return received; }

        @Override
        public <P extends Datum> void onEvent(Topic<P> topic, P payload) {
            ExtendedBar b = (ExtendedBar) payload;
            received.add(b.symb() + "|" + b.timestamp() + "|" + b.cl() + "|" + b.tradeCount());
            log.info("recv " + topic.name() + " " + b.symb() + "@" + b.timestamp()
                    + " cl=" + b.cl() + " bidVol=" + b.bidVolume() + " askVol=" + b.askVolume()
                    + " n=" + b.tradeCount());
        }
    }

    private static void awaitStarted(Gateway gateway, long timeoutMillis) throws InterruptedException {
        long deadline = System.currentTimeMillis() + timeoutMillis;
        while (gateway.status().ordinal() < GatewayStatus.STARTED.ordinal()
                && System.currentTimeMillis() < deadline) {
            Thread.sleep(5);
        }
    }

    /**
     * Assembles and runs the ExtendedBar flow. When {@code realTime} the window spans "now" so the
     * engine selects {@link OperatingMode#REAL_TIME}; otherwise the window is entirely in the past
     * and the run is a deterministic {@link OperatingMode#COMPRESSED_TIME} replay.
     *
     * @return {@code true} if the run completed and the consumer recorded exactly the expected bars
     */
    static boolean execute(boolean realTime) throws Exception {
        long start, end;
        long[] times;
        if (realTime) {
            long base = System.currentTimeMillis() + 1_500L;   // window spans now -> REAL_TIME
            times = new long[] {base, base + 800, base + 1_600};
            start = base - 500;
            end   = base + 5_000;
        } else {
            long s = 1_283_630_000_000L;                       // window in the past -> COMPRESSED_TIME
            times = new long[] {s, s + 1_000, s + 2_000};
            start = s;
            end   = s + 5_000;
        }

        Topic<ExtendedBar> bars = new Topic<>(TOPIC, ExtendedBar.class);
        MultiClientEngine engine = new MultiClientEngine("Engine", start, end);

        ExtendedBarFeed feed = new ExtendedBarFeed("bar-feed", bars, times, start, end);
        RecordingConsumer consumer = new RecordingConsumer("consumer");

        engine.registerPublisher(feed, bars, List.of(SYMB));
        engine.registerActor(consumer, Map.of(bars, List.of(SYMB)), Map.of());

        Thread engineThread = new Thread(engine, "engine");
        engineThread.start();
        awaitStarted(engine, 2_000);
        Thread feedThread = new Thread(feed, "bar-feed");
        feedThread.start();

        engineThread.join(20_000);
        feedThread.join(2_000);

        // Observer surface: what the run did, including the discovered providers. Actors never
        // see this; a strategy cannot tell replay from live.
        RunInfo info = engine.runInfo();
        List<String> received = consumer.received();
        boolean ok = engine.status() == GatewayStatus.COMPLETE && received.equals(expected(times));

        LOG.info("mode=" + info.mode()
                + " window=[" + info.startTime() + ".." + info.endTime() + "]"
                + " typeFingerprint=" + info.typeFingerprint()
                + " providers=" + info.providerIds()
                + " status=" + engine.status() + "; received " + received.size() + " bars");
        for (String r : received) {
            System.out.println("   " + r);
        }
        System.out.println(ok ? "EXTENDED BAR RUN OK" : "EXTENDED BAR RUN FAILED");
        return ok;
    }

    private ExtendedBarRun() {}
}

File: examples/pulse-ext-example/examples/extended_bar_run.py
# SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Inventzia-Commercial
# Copyright (c) 2013-2026 Magrino Bini, Paola Apruzzese, Inventzia Science and Technology Ltd.
"""Shared run logic behind the two runnable scripts, ``historical_extended_bar_run.py``
and ``realtime_extended_bar_run.py``: the extension's ExtendedBar flowing through the Java
engine (JPype).

A Python source gateway emits a short stream of ExtendedBar (the ad-hoc datum defined by
this package), the Java engine routes them, and a Python consumer receives them with all
fields intact. The extension jar is placed on the JVM classpath so Java's ServiceLoader
discovers the same ExtendedBar provider Python discovered via its entry point; the bridge's
type-universe gate then confirms both runtimes agree before any event flows.

This module holds no ``__main__``: run one of the two scripts instead. The only difference
between them is the time window, which selects compressed-time (historical replay) or
real-time (paced to the wall clock) via ``run(realtime=...)``.
"""

import threading
import time
from decimal import Decimal
from pathlib import Path

from inventzia.pulse.beacon.core.actor import BeaconActor
from inventzia.pulse.beacon.core.crosslanguage import jpype_host
from inventzia.pulse.beacon.core.crosslanguage.cross_language_streamer import CrossLanguageStreamer
from inventzia.pulse.beacon.core.gateway import BeaconGateway
from inventzia.pulse.beacon.core.reporter import ComponentReporter, ReportLevel, configure_logging
from inventzia.pulse.ext.schemas.extended_bar import ExtendedBar

LOG = ComponentReporter("extended-bar-run")

_EXT_JAR = Path(__file__).resolve().parents[1] / "java" / "target" / "pulse-ext-example-0.1.0.jar"

SYMB = "AAPL"
_BARS = [
    (Decimal("1.00"), Decimal("1.20"), Decimal("0.95"), Decimal("1.10"), 40, 60, 3),
    (Decimal("1.10"), Decimal("1.30"), Decimal("1.05"), Decimal("1.25"), 55, 45, 5),
    (Decimal("1.25"), Decimal("1.28"), Decimal("1.10"), Decimal("1.15"), 30, 70, 4),
]


def _bar(t, row):
    op, hi, lo, cl, bidv, askv, n = row
    return ExtendedBar(symb=SYMB, timestamp=t, op=op, hi=hi, lo=lo, cl=cl,
                       vlm=bidv + askv, bidVolume=bidv, askVolume=askv, tradeCount=n, venue="XNAS")


class ExtendedBarFeed(BeaconGateway):
    """Source gateway emitting a fixed stream of ExtendedBar."""

    def __init__(self, name, topic_name, times):
        super().__init__(name)
        self._topic_name = topic_name
        self._times = times

    def produce(self):
        for t, row in zip(self._times, _BARS):
            yield self._topic_name, _bar(t, row)


class RecordingConsumer(BeaconActor):
    def __init__(self, name, sink):
        super().__init__(name)
        self._sink = sink

    def on_event(self, topic_name, datum):
        self._sink.append((str(datum.datum_key), int(datum.datum_time),
                           str(datum.cl), int(datum.trade_count)))
        LOG.info(f"recv {topic_name} {datum.datum_key}@{datum.datum_time} "
                 f"cl={datum.cl} bidVol={datum.bid_volume} askVol={datum.ask_volume} n={datum.trade_count}")


def _await_started(engine, timeout_s=5.0):
    deadline = time.monotonic() + timeout_s
    while time.monotonic() < deadline:
        if str(engine.status().name()) == "STARTED":
            return
        time.sleep(0.02)
    raise TimeoutError("engine did not reach STARTED")


def expected(times):
    return [(SYMB, t, str(row[3]), row[6]) for t, row in zip(times, _BARS)]


def run(realtime=False):
    """Run the cross-language ExtendedBar demo; returns ``(received, status, times)``."""
    configure_logging(ReportLevel.INFO)
    if not _EXT_JAR.exists():
        raise FileNotFoundError(f"build the extension jar first: {_EXT_JAR} (mvn -f java/pom.xml package)")

    if realtime:
        base = int(time.time() * 1000) + 1_500          # window spans now → REAL_TIME
        times = [base, base + 800, base + 1_600]
        start, end = base - 500, base + 5_000
    else:
        start = 1_283_630_000_000                        # window in the past → COMPRESSED_TIME
        times = [start, start + 1_000, start + 2_000]
        start, end = start, start + 5_000

    jpype_host.start_jvm(extra_classpath=[_EXT_JAR])     # gate verifies core+ext == core+ext
    from jpype import JClass

    MultiClientEngine = JClass("com.inventzia.pulse.beacon.core.MultiClientEngine")
    Topic = JClass("com.inventzia.pulse.beacon.core.Topic")
    CrossLanguageActor = JClass("com.inventzia.pulse.beacon.core.crosslanguage.CrossLanguageActor")
    CrossLanguageGateway = JClass("com.inventzia.pulse.beacon.core.crosslanguage.CrossLanguageGateway")
    ExtendedBarJ = JClass("com.inventzia.pulse.ext.schemas.ExtendedBar")   # the extension's Java record
    JList = JClass("java.util.List")
    JMap = JClass("java.util.Map")
    JThread = JClass("java.lang.Thread")

    bars = Topic("ext.bars", ExtendedBarJ.class_)
    engine = MultiClientEngine("Engine", start, end)

    py_feed = CrossLanguageGateway("Py-bar-feed", start, end).withTopic(bars)
    py_consumer = CrossLanguageActor("Py-consumer")
    engine.registerPublisher(py_feed, bars, JList.of(SYMB))
    engine.registerActor(py_consumer, JMap.of(bars, JList.of(SYMB)), JMap.of())

    received = []
    consume_streamer = CrossLanguageStreamer(py_consumer, RecordingConsumer("Consumer", received))
    produce_streamer = CrossLanguageStreamer(py_feed, ExtendedBarFeed("Feed", "ext.bars", times))
    errors = {}

    def _guarded(fn, label, **kwargs):
        def wrapper():
            try:
                fn(**kwargs)
            except BaseException as exc:  # noqa: BLE001
                errors.setdefault(label, exc)
                LOG.severe(f"streamer '{label}' failed: {exc!r}")
        return wrapper

    t_consume = threading.Thread(target=_guarded(consume_streamer.run_consume, "consume"),
                                 name="consume", daemon=True)
    t_consume.start()
    engine_thread = JThread(engine, "Engine-thread")
    engine_thread.start()
    _await_started(engine)
    JThread(py_feed, "py-feed").start()
    t_produce = threading.Thread(
        target=_guarded(produce_streamer.run_produce, "produce",
                        on_error=lambda e: errors.setdefault("produce", e)),
        name="produce", daemon=True)
    t_produce.start()

    deadline = time.monotonic() + 20.0
    while engine_thread.isAlive() and time.monotonic() < deadline:
        if errors:
            break
        time.sleep(0.05)
    if errors and engine_thread.isAlive():
        engine_thread.interrupt()
        engine_thread.join(2_000)
    for t in (t_produce, t_consume):
        t.join(5.0)
    if errors:
        label, exc = next(iter(errors.items()))
        raise RuntimeError(f"extended-bar run aborted: streamer '{label}' failed") from exc

    return received, str(engine.status().name()), times


def main(realtime=False):
    """Run the flow and report; returns a process exit code (0 on success).

    Used by the two runnable scripts, ``historical_extended_bar_run.py`` and
    ``realtime_extended_bar_run.py``.
    """
    received, status, times = run(realtime=realtime)
    ok = status == "COMPLETE" and received == expected(times)
    LOG.info(f"mode={'REAL_TIME' if realtime else 'COMPRESSED'} status={status}; received {len(received)} bars")
    for r in received:
        print("  ", r)
    print("EXTENDED BAR RUN OK" if ok else "EXTENDED BAR RUN FAILED")
    return 0 if ok else 1
